CppImport/src/macro/LexicalHelper.cpp

```diff
@@ -0,0 +1,283 @@
+ **      derived from this software without specific prior written permission.
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "LexicalHelper.h"
+
+#include "ExpansionManager.h"
+#include "StaticStuff.h"
+
+namespace CppImport {
+
+LexicalHelper::LexicalHelper(ClangHelper* clang, ExpansionManager* expansionManager)
+	: clang_(clang), expansionManager_(expansionManager) {}
+
+bool LexicalHelper::isConcatenationStringifycation(clang::SourceLocation loc)
+{
+	if (loc.isMacroID())
+		if (auto immediateExpansion = expansionManager_->immediateExpansion(loc))
+			return clang_->sourceManager()->getImmediateExpansionRange(loc).first !=
+					immediateExpansion->range.getBegin();
+
+	return false;
+}
+
+clang::SourceRange LexicalHelper::unexpandedSourceRange(clang::SourceRange range)
+{
+	clang::SourceLocation start, end;
+
+	if (isConcatenationStringifycation(range.getBegin()))
+		start = clang_->sourceManager()->getImmediateExpansionRange(range.getBegin()).first;
+	else
+		start = range.getBegin();
+
+	if (isConcatenationStringifycation(range.getEnd()))
+		end = clang_->sourceManager()->getImmediateExpansionRange(range.getEnd()).second;
+	else
+		end = range.getEnd();
+
+	return clang::SourceRange(start, end);
+}
+
+QString LexicalHelper::unexpandedSpelling(clang::SourceRange range)
+{
+	auto result = clang_->spelling(unexpandedSourceRange(range));
+	while (result.startsWith("\\")) result = result.right(result.length() - 1);
+
+	return result.trimmed();
+}
+
+void LexicalHelper::correctNode(clang::Decl* clangAstNode, Model::Node* envisionAstNode)
+{
+	clang::SourceRange spellingRange = clangAstNode->getSourceRange();
+
+	if (auto fDecl = clang::dyn_cast<clang::FunctionDecl>(clangAstNode))
+	{
+		spellingRange = fDecl->getNameInfo().getSourceRange();
+
+		auto ooMethod = DCast<OOModel::Method>(envisionAstNode);
+		Q_ASSERT(ooMethod);
+
+		if (ooMethod->results()->size() > 0)
+			correctNode(fDecl->getReturnTypeSourceRange(), ooMethod->results()->first());
+
+		int i = 0;
```

Comment:
If you need an index `i` anyway, can't you just use that, instead of an iterator for the loop body?